### PR TITLE
add skipCadvisorProcessMetrics in kubelet config.yaml for tuning

### LINKED ISSUE #99183: Make kubelet cadvisor optional to read every open FDs of every container.
#### What would you like to be added:
When running `elasticsearch` workload,  it's possible that `elasticsearch` has huge number of open files. This causes `kubelet` CPU usage extremely high and impacts the application responsiveness. 

CPU profiling data shows the hot code is here: 
https://github.com/kubernetes/kubernetes/blob/release-1.18/vendor/github.com/google/cadvisor/container/libcontainer/handler.go#L163-L182
```go
	for _, pid := range pids {
		dirPath := path.Join(rootFs, "/proc", pid, "fd")
		fds, err := ioutil.ReadDir(dirPath)
		if err != nil {
			klog.V(4).Infof("error while listing directory %q to measure fd count: %v", dirPath, err)
			continue
		}
		fdCount += uint64(len(fds))
		for _, fd := range fds {
			fdPath := path.Join(dirPath, fd.Name())
			linkName, err := os.Readlink(fdPath)
			if err != nil {
				klog.V(4).Infof("error while reading %q link: %v", fdPath, err)
				continue
			}
			if strings.HasPrefix(linkName, "socket") {
				socketCount++
			}
		}
	}
``` 
Is counting files and sockets absolutely necessary for k8s to be functional? If not,  is it possible to add a kubelet flag to disable above code path? 

#### Why is this needed:
Reduce the resource usage of kubelet to make the application run well.

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?

alter proposal with #107630 （use feature gate）
/kind feature

#### What this PR does / why we need it:
Not sure if we need a KEP for this. But this is a really simple flag.

Use case from @xinfengliu:

> When running elasticsearch workload, it's possible that elasticsearch has huge number of open files. This causes kubelet CPU usage extremely high and impacts the application responsiveness.
> 
> CPU profiling data shows the hot code is here:
> https://github.com/kubernetes/kubernetes/blob/release-1.18/vendor/github.com/google/cadvisor/container/libcontainer/handler.go#L163-L182

#### Which issue(s) this PR fixes:

Fixes #99183

#### Special notes for your reviewer:
Some details about the `cadvisor.ProcessMetrics` are in #99183 (comment).

> Process metrics include container_processes, container_file_descriptors, container_sockets, container_threads_max, container_threads, container_ulimits_soft.
> 
> If I search correctly, they are not used in other APIs.

- another proposal would be adding a kubelet flag to disabling it.
- or we can edit the interval to get process metrics in cadvisor to make the performance issue not that concerning.

#### Does this PR introduce a user-facing change?
```release-note
kubelet: add skipCadvisorProcessMetrics flag to skip it for a performance issue of cadvisor
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/apis/config/types.go [hunk 1]
```diff
@@ -444,6 +444,10 @@ type KubeletConfiguration struct {
 	// registerNode enables automatic registration with the apiserver.
 	// +optional
 	RegisterNode bool
+
+	// SkipCadvisorProcessMetrics provide a way to skip the cadvisormetrics.ProcessMetrics
```
Comment (dashpole) [inline pkg/kubelet/apis/config/types.go:448]: Can we document why someone might want to skip these metrics?

Comment (dashpole) [reply pkg/kubelet/apis/config/types.go:448]: Can you document what cadvisor/summary API metrics will be missing as a result?

**Review by dashpole — COMMENTED** [review]
Can we validate that this option isn't enabled at the same time as PID eviction?

**Comment by pacoxu** [comment]
> The failures don't seem related. I don't love this solution, as i'd prefer to improve the performance of stats collection if possible, but if we need a knob to disable this behavior, i'm OK with it.
> cc @bobbypage

This should be optimized or fixed in cadvisor. 

https://github.com/kubernetes/kubernetes/blob/release-1.18/vendor/github.com/google/cadvisor/container/libcontainer/handler.go#L163-L182

May I open a cadvisor issue to track it? @bobbypage

File: pkg/kubelet/apis/config/types.go (hunk 1 above)
Comment (pacoxu) [reply pkg/kubelet/apis/config/types.go:448]: Of course. 

	// SkipCadvisorProcessMetrics provide a way to skip the cadvisormetrics.ProcessMetrics, as there is a performance issue when there is a huge number of open files. 
        // See https://github.com/kubernetes/kubernetes/issues/99183 for more details.
        // When skipping process metrics,  the cadvisor metrics would not include container_processes, container_file_descriptors, container_sockets, container_threads_max, container_threads, container_ulimits_soft. 
```

**Comment by bobbypage** [comment]
> https://github.com/kubernetes/kubernetes/blob/release-1.18/vendor/github.com/google/cadvisor/container/libcontainer/handler.go#L163-L182

Makes sense to open cAdvisor issue to track it. Is there something we can do to improve perf there? Does cAdvisor need to iterate through every fd?

**Comment by pacoxu** [comment]
I opened https://github.com/google/cadvisor/issues/3063 to track it. 

The logic was included by https://github.com/google/cadvisor/pull/2260.

**Comment by pacoxu** [comment]
According to https://github.com/google/cadvisor/issues/3063#issuecomment-1039930903, the process metrics are disabled by default in cadvisor, related flags should also be added in k8s.

Is that mean that we should remove ProcessMetrics with a feature gate? Then in a later release, we can promote it to beta and finally remove it with the feature gate.  So I prefer to process https://github.com/kubernetes/kubernetes/pull/107630/files on.

**Comment by pacoxu** [comment]
I prefer the proposal in https://github.com/kubernetes/kubernetes/pull/107630 or fix it in cadvisor.

**Pull request closed without merging**